hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java

```diff
@@ -1025,4 +1026,27 @@ private static void checkNotPagingPredicate(Predicate predicate, String method)
             throw new IllegalArgumentException("PagingPredicate not supported in " + method + " method");
         }
     }
+
+    @Override
+    public V computeIfPresent(K key,
+                               BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
+        checkNotNull(key, NULL_KEY_IS_NOT_ALLOWED);
+
+        while (true) {
+            Data oldValueData = toData(getInternal(key));
+            if (oldValueData == null) {
+                return null;
+            }
+
+            V oldValueClone = toObject(oldValueData);
+            V newValue = remappingFunction.apply(key, oldValueClone);
+            if (newValue != null) {
+                if (replaceInternal(key, oldValueData, toData(newValue))) {
+                    return newValue;
+                }
+            } else if (removeInternal(key, oldValueData)) {
+                return null;
+            }
+        }
+    }
```

Comment:
I could not think of a way to combine the get() + replace() operation into a single remote call. Mainly because I don't know how to send the BiFunction code/body across the wire (the function may've used classes/utilities that are only present on the client JVM)

Reply:
I think this is main issue that needs to be resolved. The problem is that this implementation can consume a lot of resources because of the retrying and the long latency. It could even end up in a livelock.

The best approach is to get everything done in a single remote call. There is no retrying needed here because the operation will have exclusive access to the partition.

Reply:
Sure @pveentjer , I'd look into that then. 
Just to confirm - this means that we want the BiFunction also to execute on the remote machine, correct (because we need to know the existing value of the entry before we can apply the function)?

Reply:
Yes, this probably will be the better approach. You're going to have to create a new operation which will contain the mapping function - similar to how the `EntryProcessor` is transmitted when you invoke `IMap#executeOnKey`. On that note, would it not be simpler if you wrapped the mapping function in an `EntryProcessor` and used the same methods that that method already uses?

Also, I don't see any client-side support (`ClientMapProxy`) or client tests. Please add those.

Reply:
Sounds good, thanks @mmedenjak , will do.

Reply:
@pveentjer @mmedenjak @taburet 
I tested scenarios where a client node (JVM A) would pass an `EntryProcessor` as an anonymous class or a lambda to a server node (JVM B). I can confirm that such calls fail on the sever node, because the server fails to deserialize the anonymous-class/lambda instance (the server node doesn't have that "class" on its classpath).

This may not be so much of a problem for `EntryProcessor`s (because people must be creating a named subclass file in advance, and must be deploying the subclass on both (client and server) JVMs.

However, when people call methods that take Java 8 functional classes as args, its very likely that they'd use a lambda (I've almost never seen anyone creating named subclass files for a `BiFunction` or a `Function`). In such cases, the `computeIfPresent()` will fail (since the server would fail to deserialize the lambda passed as the `BiFunction`).

Given this, should we continue evaluating the `BiFunction` on the client JVM itself? 
Otherwise the new/updated implementation of `computeIfPresent ()` would cause existing usages of the method to start failing (if someone is passing the `BiFunction` arg as a lambda)

Thoughts?
PS: Should I be discussing such things on Gitter instead?

Reply:
How about if you enable user code deployment? Does that work with lambdas? For 4.1, there's some work in progress to enable re-deployment of such classes using UCD.

Reply:
Hi @cangencer, great suggestion! I tested that if we enable UCD on both sides, and add all the client-only classes to the UCD config, then that works! 
There are some nuances to it, but I think I'd raise a PR using UCD and see where that goes.

Reply:
You're right and there is another issue that we might have that we have thought of - classes might be defined on all instances but since `BiFunction` doesn't implement `Serializable` or any other serialization technique, it might not work until you either cast it or you define a named class with some serialisation technique defined. Unfortunately, UCD doesn't help there.

Feels like we should try to get the EP approach running but if it's too hacky, we might revert back to the original approach from running from the client. Yes, it has its' drawbacks but it's better than an infinite loop.

Reply:
Sure @mmedenjak - let me send through a PR with this approach (using EP + UCD) - like you said, we can take a decision then ...